torch/csrc/jit/passes/inline_forked_closures.cpp

```diff
@@ -58,6 +58,56 @@ void inlineForkedClosure(Node* fork_closure, NodeKind genKind) {
   runCleanupPasses(fork_graph);
 }
 
+void inlineAwaitableThenClosure(Node* then_closure) {
+  Node* function_context_node = then_closure->inputs()[0]->node();
+
+  if (function_context_node->inputs().size() != 2 ||
+      function_context_node->inputs().at(0)->node()->kind() != prim::Closure ||
+      function_context_node->inputs().at(1)->node()->kind() !=
+          prim::TupleConstruct) {
+    throw ErrorReport(then_closure->sourceRange()) << "Cannot fork this value";
+  }
+
+  Node* function = function_context_node->inputs().at(0)->node();
+  Node* context = function_context_node->inputs().at(1)->node();
+  auto then_graph = function->g(attr::Subgraph)->copy();
+  auto g = then_closure->owningGraph();
+  Node* then_node = g->create(prim::awaitable_then, 0)
+                        ->insertAfter(then_closure)
+                        ->setSourceRange(then_closure->sourceRange());
+  Node* then_input_node = g->create(prim::awaitable_then_input, 1)
+                              ->insertBefore(then_node)
+                              ->setSourceRange(then_closure->sourceRange());
+  TORCH_INTERNAL_ASSERT(then_closure->inputs().size() == 2);
+  auto aw = then_closure->inputs().at(1);
+  TORCH_INTERNAL_ASSERT(aw->type()->kind() == AwaitType::Kind);
+  auto fake_aw_then_input = then_input_node->output();
+  then_input_node->addInput(aw);
+
+  then_node->addInput(aw);
+  then_node->addInput(fake_aw_then_input);
+
+  auto then_graph_context = then_graph->inputs().at(0);
```

Comment:
can you add a comment describing that this is just unpacking the context tuple into individual inputs.